Invalidate :default pseudo-class changes

heycam opened the pull request:
#### ec255a8cd5cce23106c903bd2fc8620c7eac1e01
<pre>
Invalidate :default pseudo-class changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=261382">https://bugs.webkit.org/show_bug.cgi?id=261382</a>
&lt;rdar://problem/115236525&gt;

Reviewed by Tim Nguyen.

On an &lt;input type=checked&gt;, the presence of the checked=&quot;&quot; attribute
determines whether :default matches. We must invalidate :default when
the attribute is added or removed on a checkable input element.

We also must invalidate :default when the type=&quot;&quot; changes. Just use
AnyValue rather than compute ahead of time what the new pseudo-class
match value will be, in HTMLInputElement::updateType.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/input-pseudo-classes-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/input-pseudo-classes-in-has.html:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::matchesDefaultPseudoClass const):
(WebCore::HTMLInputElement::setDefaultCheckedState):
(WebCore::HTMLInputElement::copyNonAttributePropertiesFromElement):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: <a href="https://commits.webkit.org/267848@main">https://commits.webkit.org/267848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d262947bdf1f6d19b25657607435560ced0a844c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18205 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18761 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19706 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16725 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18359 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18094 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15541 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20574 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16294 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22831 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16463 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20696 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14422 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16135 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20488 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2192 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -780,9 +782,12 @@ void HTMLInputElement::attributeChanged(const QualifiedName& name, const AtomStr
+    case AttributeNames::checkedAttr: {
+        std::optional<Style::PseudoClassChangeInvalidation> checkedInvalidation;
+        if (m_inputType->isCheckable()) {
             invalidateStyleForSubtree();
```
If PseudoClassChangeInvalidation is sufficient, `invalidateStyleForSubtree` can be removed.

heycam replied:
My change here was not necessary in the first place; we already invalidate PseudoClassType::Checked in setChecked.

nt1m commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -780,9 +782,10 @@ void HTMLInputElement::attributeChanged(const QualifiedName& name, const AtomStr
+    case AttributeNames::checkedAttr: {
+        std::optional<Style::PseudoClassChangeInvalidation> checkedInvalidation;
         if (m_inputType->isCheckable())
-            invalidateStyleForSubtree();
```
The test failure suggests that we also need to use `PseudoClassChangeInvalidation` for the default pseudo class when the checked attribute changes (there's probably a test lacking for input[type=radio] + checked attribute + :has(:default), that would show the bug)

heycam replied:
Yes, that is right.

nt1m commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1000,6 +1001,17 @@ bool HTMLInputElement::isTextType() const
     return m_inputType->isTextType();
 }
 
+void HTMLInputElement::setDefaultCheckedState(bool isDefaultChecked)
+{
+    if (m_isDefaultChecked == isDefaultChecked)
+        return;
+
+    std::optional<Style::PseudoClassChangeInvalidation> defaultInvalidation;
+    if (m_inputType->isCheckable())
+        defaultInvalidation.emplace(*this, CSSSelector::PseudoClassType::Default, isDefaultChecked);
+    m_isDefaultChecked = isDefaultChecked;
+}
```
Can we add a wpt for this case involving :has()? Expanding the existing one should be fine.

heycam replied:
https://github.com/web-platform-tests/wpt/pull/41903

nt1m commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -512,6 +512,8 @@ void HTMLInputElement::updateType(const AtomString& typeAttributeValue)
         return;
     ASSERT(m_inputType->type() != newType->type());
 
+    Style::PseudoClassChangeInvalidation defaultInvalidation(*this, CSSSelector::PseudoClassType::Default, Style::PseudoClassChangeInvalidation::AnyValue);
```
I think you can now do more fine grained invalidation than AnyValue with the new m_isDefaultChecked flag.

heycam replied:
:default can also match some other input types, and I didn't want to add extra logic for those. I guess I could've done the more fine grained invalidation by checking the specific types we're changing from / to but it didn't seem worth it.

heycam committed:
Invalidate :default pseudo-class changes
https://bugs.webkit.org/show_bug.cgi?id=261382
<rdar://problem/115236525>

Reviewed by Tim Nguyen.

On an <input type=checked>, the presence of the checked="" attribute
determines whether :default matches. We must invalidate :default when
the attribute is added or removed on a checkable input element.

We also must invalidate :default when the type="" changes. Just use
AnyValue rather than compute ahead of time what the new pseudo-class
match value will be, in HTMLInputElement::updateType.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/input-pseudo-classes-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/input-pseudo-classes-in-has.html:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::matchesDefaultPseudoClass const):
(WebCore::HTMLInputElement::setDefaultCheckedState):
(WebCore::HTMLInputElement::copyNonAttributePropertiesFromElement):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: https://commits.webkit.org/267848@main
